core/sched.c

```diff
@@ -126,38 +125,43 @@ int __attribute__((used)) sched_run(void)
                        : active_thread->pid),
         next_thread->pid);
 
-    if (active_thread == next_thread) {
+    next_thread->status = STATUS_RUNNING;
+
+    if (previous_thread == next_thread) {
+#ifdef MODULE_SCHED_CB
```

Comment:
why this?

Reply:
I meant the sched_cb call only, if there's no change, didn't you write in the description that it shouldn't count as context switch?

Reply:
Assuming this is about the sched_cb, This must be called when there is a sleep between running the same thread. Otherwise a module such as `schedstatistics` could get skewed statistics when there is a sleep in the middle of thread execution.

So a context switch to the same thread, like what happens in `tests/bench_sched_nop` doesn't trigger the sched_cb call, but a thread that yields to sleep and is woken up again should trigger this to not count the time sleeping as that thread being actively scheduled.

Reply:
The situations are:

1. *Yield to same thread*: doesn't call sched_cb
2. *Yield, sleep, then back to same thread*: calls sched_cb twice
3. *Yield to different thread*: calls sched_cb twice
4. *Yield, sleep, then schedule different thread*: calls sched_cb twice

We could try to optimize case 3 and call sched_cb only once there.

Reply:
Nice explanation, mind adding this as a comment?

Reply:
Added some explanatory comments.